test/Microsoft.Azure.Devices.Edge.Test.Common/config/EdgeConfigBuilder.cs

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Test.Common.Config
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using Microsoft.Azure.Devices.Edge.Util;
+
+    public class EdgeConfigBuilder
+    {
+        readonly string deviceId;
+        readonly Dictionary<string, IModuleConfigBuilder> moduleBuilders;
+        Option<(string address, string username, string password)> registry;
+
+        public EdgeConfigBuilder(string deviceId)
+        {
+            this.deviceId = deviceId;
+            this.moduleBuilders = new Dictionary<string, IModuleConfigBuilder>();
+            this.registry = Option.None<(string, string, string)>();
+        }
+
+        public void AddRegistryCredentials(string address, string username, string password)
+        {
+            this.registry = Option.Some((address, username, password));
+        }
+
+        public IModuleConfigBuilder AddEdgeAgent(string image = null)
+        {
+            Option<string> imageOption = Option.Maybe(image);
+            var builder = new AgentModuleConfigBuilder(imageOption);
+            this.moduleBuilders.Add(builder.Name, builder);
+            return builder;
+        }
+
+        public IModuleConfigBuilder AddEdgeHub(string image = null)
+        {
+            Option<string> imageOption = Option.Maybe(image);
+            var builder = new HubModuleConfigBuilder(imageOption);
+            this.moduleBuilders.Add(builder.Name, builder);
+            return builder;
+        }
+
+        public IModuleConfigBuilder AddModule(string name, string image)
+        {
+            var builder = new ModuleConfigBuilder(name, image);
+            this.moduleBuilders.Add(builder.Name, builder);
+            return builder;
+        }
+
+        public EdgeConfiguration Build()
+        {
+            // Build all modules *except* edge agent
+            List<ModuleConfiguration> modules = this.moduleBuilders
+                .Where(b => b.Key != "$edgeAgent")
+                .Select(b => b.Value.Build())
+                .ToList();
+
+            // Build edge agent
+            modules.Insert(0, this.BuildEdgeAgent(modules));
+
+            // Compose edge configuration
+            var config = new ConfigurationContent
+            {
+                ModulesContent = new Dictionary<string, IDictionary<string, object>>()
+            };
+
+            var moduleNames = new List<string>();
+
+            foreach (ModuleConfiguration module in modules)
+            {
+                moduleNames.Add(module.Name);
+
+                if (module.DesiredProperties.Count != 0)
+                {
+                    config.ModulesContent[module.Name] = new Dictionary<string, object>
+                    {
+                        ["properties.desired"] = module.DesiredProperties
+                    };
+                }
+            }
+
+            return new EdgeConfiguration(this.deviceId, moduleNames, config);
```

Comment:
should it be better to pass modules (List<ModuleConfiguration>) to EdgeConfiguration? so we don't need to worry about number of modulesNames doesn't match config (ConfigurationContent)?

Reply:
I ended up adding the `moduleNames` parameter just so I can print out the names in the logs after we deploy the config to IoT Hub--it's not used for anything critical. I'm not super happy about this solution, but it was _way_ easier than parsing the names back out of the config. And the builder is the one calling this constructor, not the user.